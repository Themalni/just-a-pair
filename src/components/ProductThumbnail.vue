<template>
    <div class="col-md-4 col-sm-6">
        <router-link :to="'/product-view/' + id" class="product-link">
            <div class="product-thumbnail pb-2" @click="showProduct(id)">
                <div v-if="!onSale" v-show="offer" class="offer-tag" :class="{accent: onSale}">{{ offer }}</div>
                <div v-else v-show="offer" class="offer-tag" :class="{accent: onSale}">{{ salePercentage + "%" }}</div>
                <div class="product-image">
                    <img :src="image">
                </div>
                <h3 class="fancy-font mt-2 mb-2">{{ title }}</h3>
                <span class="product-price fancy-font" :class="{ 'sale-price': onSale }">${{ price }}</span>
                <span v-if="onSale" class="product-price fancy-font deep-orange-text">${{ price | salePrice(salePercentage) }}</span>
            </div>
        </router-link>
    </div>
</template>

<script>

export default {
    name: 'product-thumbnail',
    props: [
        'id',
        'image', 
        'offer', 
        'title', 
        'price', 
        'onSale', 
        'salePercentage', 
        'stock', 
        'sliderImages'
    ],
    computed: {
        products() { return this.$store.state.products; }
    },
    methods: {
        showProduct(id) {
            this.$store.dispatch('showProduct', id);
        }
    } 
}
</script>

<style lang="scss" scoped>
@import "../styles/scss/styles.scss";

.product-thumbnail {
    width: 100%;
    margin-bottom: 1.5em;
    background-color: $light-grey;
    position: relative;
    color: $dark-grey;

    &:hover {

        .product-image {
            img {
                transform: scale(1.1);
                transition: all 0.5s linear;
            }
        }
    }
}
.offer-tag {
    width: 40px;
    height: 40px;
    background-color: $main-blue;
    color: #fff;
    line-height: 2.8;
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
}
.product-image {
    width: 100%;
    background-color: $light-grey;
    overflow: hidden;
    
    img {
        transition: all 0.5s linear;
    }
}
.product-price {
    color: grey;
    font-size: 2em;
}
.product-link {
    &:hover {
        text-decoration: none;
    }
}
.accent {
    background-color: $orange-accent;
}
.btn {
    margin: 0.5em auto;
    display: block;
}
.sale-price {
    text-decoration: line-through;
    margin: 0 0.5em;
}
</style>