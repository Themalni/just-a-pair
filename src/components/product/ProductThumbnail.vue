<template>
  <div class="col-md-4 col-sm-6">
    <router-link :to="'/product/' + product.id" class="product-link">
      <div class="product-thumbnail pb-2" @click="showProduct()">
        <div v-if="!product.sale" class="offer-tag">{{ product.offer }}</div>
        <div
          v-else
          class="offer-tag"
          :class="{accent: product.sale}"
        >{{ `-${product.salePercentage}%` }}</div>
        <div class="product-image">
          <img :src="product.image">
        </div>
        <h3 class="fancy-font mt-2 mb-2">{{ product.title }}</h3>
        <span
          class="product-price fancy-font"
          :class="{ 'sale-price': product.sale }"
        >${{ product.price }}</span>
        <span
          v-if="product.sale"
          class="product-price fancy-font deep-orange-text"
        >${{ product.price | salePrice(product.salePercentage) }}</span>
      </div>
    </router-link>
  </div>
</template>

<script>
import { mapGetters, mapMutations, mapActions } from "vuex";

export default {
  name: "ProductThumbnail",
  props: {
    product: {
      type: Object
    }
  },
  computed: {
    ...mapGetters(["productPreview"])
  },
  methods: {
    ...mapActions(["clearProductPreview"]),
    showProduct() {
      this.clearProductPreview();
      this.$store.dispatch("showProduct", this.product);
    }
  }
};
</script>

<style lang="scss" scoped>
.product-thumbnail {
  width: 100%;
  margin-bottom: 1.5em;
  background-color: $light-grey;
  position: relative;
  color: $dark-grey;

  &:hover {
    .product-image {
      img {
        transform: scale(1.1);
        transition: all 0.5s linear;
      }
    }
  }
}
.offer-tag {
  width: 40px;
  height: 40px;
  background-color: $main-blue;
  color: #fff;
  line-height: 2.8;
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
}
.product-image {
  width: 100%;
  background-color: $light-grey;
  overflow: hidden;

  img {
    transition: all 0.5s linear;
  }
}
.product-price {
  color: grey;
  font-size: 2em;
}
.product-link {
  &:hover {
    text-decoration: none;
  }
}
.accent {
  background-color: $orange-accent;
}
.btn {
  margin: 0.5em auto;
  display: block;
}
.sale-price {
  text-decoration: line-through;
  margin: 0 0.5em;
}
</style>
