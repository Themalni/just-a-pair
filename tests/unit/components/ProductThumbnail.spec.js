import { mount, createLocalVue } from '@vue/test-utils';
import ProductThumbnail from '@/components/Product/ProductThumbnail';
import mockState from '../utils/mockState';
import Vuex from 'vuex';
import salePrice from "@/filters.js";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.filter('salePrice', salePrice);

describe('ProductThumbnail', () => {
    let wrapper
    let store
    let state

    beforeEach(() => {
        state = mockState.state
        store = new Vuex.Store({
            state
        })

        wrapper = mount(ProductThumbnail, {
            stubs: ['router-link'],
            store,
            localVue,
            propsData: {
                product: {
                    id: 621,
                    title: 'Summer sandals',
                    sku: 'BR55-36-Black',
                    price: 120,
                    salePercentage: 30,
                    color: 'Black',
                    image: '../assets/sneakers.jpg',
                    sliderImages: [
                        '../assets/sneakers.jpg',
                        '../assets/sneakers.jpg',
                        '../assets/sneakers.jpg'
                    ],
                    sale: true,
                    amount: 1,
                    size: 37
                }
            }
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('`SET_PRODUCT_TO_PREVIEW` is dispatched when thumbnail is clicked', async () => {
        const showProductMock = jest.spyOn(wrapper.vm, 'showProduct');

        await wrapper.find(".product-thumbnail").trigger("click")

        expect(showProductMock).toBeCalled();
    });
})