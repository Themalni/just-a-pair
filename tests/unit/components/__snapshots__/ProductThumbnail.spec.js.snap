// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductThumbnail renders correctrly 1`] = `
<div class="col-md-4 col-sm-6">
  <router-link-stub to="/product/621" class="product-link">
    <div class="product-thumbnail pb-2">
      <div class="offer-tag accent">-30%</div>
      <div class="product-image"><img src="../assets/sneakers.jpg"></div>
      <h3 class="fancy-font mt-2 mb-2">Summer sandals</h3> <span class="product-price fancy-font sale-price">
        $120
      </span> <span class="product-price fancy-font deep-orange-text">
        $84
      </span>
    </div>
  </router-link-stub>
</div>
`;
